torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -396,18 +458,168 @@ class TensorExprFuser {
+
+      for (Block* b : n->blocks()) {
+        fillTypeInfoMap(b);
+      }
+    }
+  }
+  void printTypeInfoMap() {
+    GRAPH_DEBUG("Typeinfo map:");
+    for (const auto& kv : typeinfo_map_) {
+      // GRAPH_DEBUG("%", kv.first->debugName(), " --> ", *kv.second);
+      GRAPH_DEBUG("%??? --> ", *kv.second);
+    }
+  }
+
+  void removeProfilingNodes(Block* b) {
+    for (auto it = b->nodes().begin(); it != b->nodes().end(); it++) {
+      if (it->kind() == prim::profile) {
+        GRAPH_DEBUG("TYPEINFO ERASE: %", it->output()->debugName());
+        typeinfo_map_.erase(it->output());
+        it->output()->replaceAllUsesWith(it->input());
+        it.destroyCurrent();
+      } else {
+        for (Block* ib : it->blocks()) {
+          removeProfilingNodes(ib);
+        }
+      }
+    }
+  }
+
+  void guardFusionGroup(Node* fusion_group) {
+    GRAPH_DEBUG("Inserting a typecheck guard for a node", *fusion_group);
+    auto subgraph = SubgraphUtils::getSubgraph(fusion_group);
+
+    // Fixup types of the subgraph inputs
+    std::vector<Value*> inputs_to_check;
+    for (size_t idx = 0; idx < subgraph->inputs().size(); idx++) {
+      // TODO: If a type of the input was explicitly set in IR (in constrast to
+      // us expecting it from a hint given by a prim::profile node), we don't
+      // need to check that input in the typecheck.
+
+      // We only check inputs of the fusion group and expect NNC to infer
+      // intermediates and outputs shapes
+      if (typeinfo_map_.count(subgraph->inputs()[idx])) {
+        subgraph->inputs()[idx]->setType(
+            typeinfo_map_.at(subgraph->inputs()[idx]));
+        inputs_to_check.push_back(fusion_group->input(idx));
+      }
+    }
+    if (!inputs_to_check.size()) {
+      return;
+    }
+
+    // Add prim::TypeCheck node
+    //
+    // TypeCheck nodes  look like the following:
+    //   %out1 : Float(2, 3), %out2 : Int(10, 30), %types_match : bool =
+    //   prim::TypeCheck(%inp1 : Tensor, %inp2 : Tensor)
+    //
+    // They have N inputs whose types we are going to check and N+1 outputs. The
+    // first N outputs specify expected types and N+1-th output holds the result
+    // of the check (bool).
+    Node* typecheck_node =
+        fusion_group->owningGraph()
+            ->create(
+                prim::TypeCheck, inputs_to_check, inputs_to_check.size() + 1)
+            ->insertBefore(fusion_group);
+    Value* typecheck_result = typecheck_node->output(inputs_to_check.size());
+
+    // Fixup types of the typecheck node outputs
+    std::unordered_map<Value*, Value*> typechecked_inputs;
+    size_t output_idx = 0;
+    for (size_t idx = 0; idx < subgraph->inputs().size(); idx++) {
+      if (typeinfo_map_.count(subgraph->inputs()[idx])) {
+        typechecked_inputs[fusion_group->input(idx)] =
+            typecheck_node->output(output_idx);
+        typecheck_node->output(output_idx++)
+            ->setType(typeinfo_map_.at(subgraph->inputs()[idx]));
+      }
+    }
```

Comment:
can we consider adding an assertion here in the else clause that `subgraph->inputs()[idx]->isCompleteTensor()` if it's of a tensor type?